File: leetcode/typescript/000322.coin-change.ts
// Time complexity: O(n)
// Space complexity: O(n)
function coinChange(coins: number[], amount: number): number {
  const dp = [0]; // represents least coin needed for i amount
  for (let i = 1; i <= amount; i++) {
    dp[i] = Infinity;
    for (const coin of coins) {
      if (coin > i) continue;
      dp[i] = Math.min(dp[i], 1 + dp[i - coin]);
    }
  }
  return dp[amount] === Infinity ? -1 : dp[amount];
}

File: leetcode/python/000322.coin-change.py
class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def coinChange(self, coins: list[int], amount: int) -> int:
        dp = [0] + [float('inf')] * amount

        for i in range(1, amount + 1):
            for coin in coins:
                if coin <= i:
                    dp[i] = min(dp[i], 1 + dp[i - coin])

        return -1 if dp[amount] == float('inf') else dp[amount]
